### PR TITLE
Implement tmt clean

### PR DESCRIPTION
**Pull request by FrNecas**
Just an initial draft for command-line interface putting together everything that was mentioned in https://github.com/psss/tmt/issues/436 to give some space for feedback/discussion. @psss @thrix does this look reasonable to you or is there anything missing or something could be improved? I will slowly start working on the implementation so we should hopefully have something working soon :)

I thought about `tmt clean images` and for now it only make sense for testcloud provision but perhaps in the future, there may be more.

Resolves: #436

### CODE REVIEW

**Comment by lukaszachy** [comment]
Early feedback: ```$ tmt clean --dry -vvv``` with outcome `removing everything` isn't very helpful. I'd prefer to see what will be done so I am sure I don't mess up things

**Comment by FrNecas** [comment]
Yeah this is a very rough first draft with no implementation and just options/help messages (I have something already implemented, will hopefully push soon). I was wondering what the expected output is when `--dry` is used. If I, as a user, saw something like

```
Removing workdir /var/tmp/tmt/run-001
Removing workdir /var/tmp/tmt/run-002
```

I would probably get scared that it actually removed the workdirs so the output should be different. Overall, dry mode in tmt is a bit incosistent, sometimes even broken so we should agree on how it should behave and what the output should be ;) What do you think? I'd propose the above for non-dry mode and then for dry mode:

```
Would remove /var/tmp/tmt/run-001
...
```

**Comment by lukaszachy** [comment]
Well, having --dry consistent across all tmt would be very nice.
Personally I like true messages as  `Would remove` more even though it is more work.
Some tools try to avoid this need by printing something among lines of 'this is a dry run, no real action will be made' but it is easy to overlook.

**Comment by FrNecas** [comment]
I just pushed a few commits with implementation, everything from the CLI should be covered, I'd be glad for any feedback/bug reports. Will start working on test coverage for this soon.

One thing that I'd like to improve (this also applies to `tmt status`) is speed... Loading runs takes time and we can't change that but making deepcopies of context slows down the commands significantly and I think it could be avoided (I will have to look into it more).

File: tmt/steps/provision/testcloud.py
Comment (FrNecas) [inline tmt/steps/provision/testcloud.py:239]: I don't really like passing `clean` and `dry` to this method but I haven't found a way to get option/log inside a classmethod. @psss am I missing a way of doing that? I think this method has to be a class method because it doesn't make much sense to instantiate an instance.

**Comment by FrNecas** [comment]
I noticed that running `tmt clean guests runs images` doesn't work because click takes `runs` as a `path` argument to `guests`. Will have to see if there is a way to workaround this behaviour...

**Comment by FrNecas** [comment]
Test coverage added, the issue with click fixed by removing chaining (it wouldn't be very useful anyway IMHO), ready for reviews/feedback.

**Comment by lukaszachy** [comment]
Thanks for the clean command, this happened when I tried it:
### 1. Domain already removed
E.g using virt-manager (or gnome-boxes for users session). 
Currently `tmt clean ` fails with
```tmt cleanup
    Guest cleanup
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'JApGSsSmHlKXHJNf'
Failed to stop testcloud instance: Instance doesn't exist: JApGSsSmHlKXHJNf
```
Note it is not clear which run it is, one has to run in verbose mode.

### 2. Removing stopped containers
After reboot container doesn't need to run and /var/tmp/tmt stays between reboots.
```
stderr (1/1 lines)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: can only stop created or running containers. e188ada784fdac73d47f56d20f87edb70ce4a87f303d8db7bb6786dc70c57473 is in state configured: container state improper

Failed to run command 'podman container stop tmt-var-tmp-tmt-run-034-plans-default'.
```
Note that `podman rm e188ada784fdac73d47f56d20f87edb70ce4a87f303d8db7bb6786dc70c57473` worked fine.

### 3. Nonexistent workdir
E.g. playing with plans/tests in /tmp. I see warn message, but imho it shouldn't matter - guest information is in provision data.
```
warn: Failed to check /var/tmp/tmt/run-039 (Invalid yaml syntax: Invalid directory path: /tmp/a).
```

### 4. Fail to remove is treated as warn
Exit code is 0 but /var/tmp/run is not empty (removal was not possible due #502)
```
tmt cleanup
    Guest cleanup
    Run cleanup
        warn: Failed to remove /var/tmp/tmt/run-001: [Errno 13] Permission denied: 'etc'
        warn: Failed to remove /var/tmp/tmt/run-003: [Errno 13] Permission denied: 'etc'
    Image cleanup
        testcloud images
```

**Comment by FrNecas** [comment]
Thanks for the feedback @lukaszachy , really appreciated. I've managed to fix 1) and 4) so far.

> Note it is not clear which run it is, one has to run in verbose mode.

Do you think that the current processed run should be displayed even in non-verbose mode? Looking at it now, it would probably make sense as there is almost no output without it and cleaning up guests may take some time so the user should know about the progress of the cleanup. So I propose changing `verbose` to `info` in all messages from this PR...

As for 2), I wasn't able to reproduce it. Looks like the container is stopped and tmt tries to run `podman stop` on the container. I've tried this locally but the cleanup passed:
```
tmt run provision -h container
podman stop tmt-var-tmp-tmt-run-001-test
tmt clean guests
```

Could you provide some more information regarding this point (e.g. how to reproduce)?

3) is going to be a bit tricky, the issue comes from `Run.load()` (https://github.com/psss/tmt/blob/master/tmt/base.py#L1013 to be exact). It tries to load a tree of a directory that apparently has been deleted (temporary directory). But you are correct that it shouldn't be necessary (the plan and provision + guest data should be stored in the run directory which exists). I will try to implement a "lite" version of run that doesn't load the tree and see how that goes.

**Comment by lukaszachy** [comment]
Thank you, I'll try it soon but I'm not able to reproduce podman problem either so it will need some time.
I run into another issue - `'/var/tmp/tmt/testcloud/images` nor  `/var/tmp/tmt/` are guaranteed to exist on the system  - If you `tmt clean` without them you end up with `FileNotFoundError`

**Comment by FrNecas** [comment]
I've just fixed the non-existent directories and also the problem with loading runs where root has been deleted. This also fixed the workaround for `tmt status` where a deepcopy of `_context.obj` was made after each run to avoid modification as we can now safely override the steps...

Moreover, the changes made `tmt status` and `tmt clean` significantly faster. Status is now almost instant even though it used to take quite a long time to process even a handful of runs :-) If you discover any side-effects of the changes, please let me know but from my observations (reproducing the case which didn't work without a deepcopy), it should work correctly.

Sorry for the number of commits, I will keep them as they are for now and rebase/reorganize/squash once the PR is approved

**Comment by lukaszachy** [comment]
Hm, I'm not sure what is wrong but I susspect qemu session change.
```
$ virsh list --all
 Id   Name               State
-----------------------------------
 2    HOEmtxmiYxssWzxR   running
 -    ElLSBeZAGFYNjfqi   shut off
 -    JkWdpfcFlhaEdPFI   shut off
 -    SDyZjzRVesYGmIzG   shut off
 -    uNGYHEtCgiUWpdGy   shut off
```
But `tmt clean guests` ends as
```tmt cleanup
    Guest cleanup
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'yKhPrkikyNnjpTIW'
        warn: Could not stop guest in run /var/tmp/tmt/run-007: Failed to stop testcloud instance: Instance doesn't exist: yKhPrkikyNnjpTIW
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'uNGYHEtCgiUWpdGy'
        warn: Could not stop guest in run /var/tmp/tmt/run-028: Failed to stop testcloud instance: Instance doesn't exist: uNGYHEtCgiUWpdGy
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'yvEJSYKPqcLfRAlw'
        warn: Could not stop guest in run /var/tmp/tmt/run-008: Failed to stop testcloud instance: Instance doesn't exist: yvEJSYKPqcLfRAlw
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'QxXkCXaOCuSBpFtq'
        warn: Could not stop guest in run /var/tmp/tmt/run-010: Failed to stop testcloud instance: Instance doesn't exist: QxXkCXaOCuSBpFtq
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'ElLSBeZAGFYNjfqi'
        warn: Could not stop guest in run /var/tmp/tmt/run-011: Failed to stop testcloud instance: Instance doesn't exist: ElLSBeZAGFYNjfqi
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'SDyZjzRVesYGmIzG'
        warn: Could not stop guest in run /var/tmp/tmt/run-029: Failed to stop testcloud instance: Instance doesn't exist: SDyZjzRVesYGmIzG
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'JkWdpfcFlhaEdPFI'
        warn: Could not stop guest in run /var/tmp/tmt/run-027: Failed to stop testcloud instance: Instance doesn't exist: JkWdpfcFlhaEdPFI
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'AOqjtnRxbeCOnata'
        warn: Could not stop guest in run /var/tmp/tmt/run-003: Failed to stop testcloud instance: Instance doesn't exist: AOqjtnRxbeCOnata
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'HOEmtxmiYxssWzxR'
        warn: Could not stop guest in run /var/tmp/tmt/run-032: Failed to stop testcloud instance: Instance doesn't exist: HOEmtxmiYxssWzxR
libvirt: QEMU Driver error : Domain not found: no domain with matching name 'fAlbqyUaLzLmjkgF'
        warn: Could not stop guest in run /var/tmp/tmt/run-009: Failed to stop testcloud instance: Instance doesn't exist: fAlbqyUaLzLmjkgF


```

**Comment by lukaszachy** [comment]
Hm, I killed tmt run while it was runnning so tmt status now says
```
disc prov prep exec repo fini  id
todo done todo todo todo done  /var/tmp/tmt/run-039  /plans/features/advanced
todo todo todo todo todo done  /var/tmp/tmt/run-039  /plans/features/basic
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-039  /plans/features/core
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-039  /plans/integration
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-039  /tests/p
```

Could `tmt clean` somehow take care of cleaning also this kind of runs?

**Comment by lukaszachy** [comment]
Could runs' `path` be an option? See
```
$ tmt clean runs guests
tmt cleanup
Path guests doesn't exist.
```

**Comment by FrNecas** [comment]
@lukaszachy what do you mean by taking care of interrupted runs? I've tested a similar scenario and it seems to work correctly, guests are stopped in the run even if not all plans have been executed and the directory is deleted:
```
$ tmt status -vv
disc prov prep exec repo fini  id
done done todo todo todo todo  /var/tmp/tmt/run-002  /plans
done done done todo todo todo  /var/tmp/tmt/run-001  /plans/features/advanced
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/features/basic
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/features/core
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/install/docs
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/install/minimal
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/install/pip
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/integration
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/provision/connect
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/provision/container
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/provision/minute
---- ---- ---- ---- ---- ----  /var/tmp/tmt/run-001  /plans/provision/virtual
$ tmt clean -v --dry
tmt cleanup
    Guest cleanup
        Would stop guests in run /var/tmp/tmt/run-002 plan /plans
        Would stop guests in run /var/tmp/tmt/run-001 plan /plans/features/advanced
    Run cleanup
        Would remove workdir /var/tmp/tmt/run-002
        Would remove workdir /var/tmp/tmt/run-001
    Image cleanup
        testcloud images
            warn: Directory /var/tmp/tmt/testcloud/images does not exist
```

As for the option, that's definitely up for some discussion. I've found out that there is no way of chaining subcommands and having arguments (this makes sense). Some time ago, I discussed this with @psss and we agreed that removing chaining makes the most sense, the reasoning being:
1) we would need to change this in tmt status for consistency as well (not a big problem)
2) hard to name the option... `--root` would make the most sense but that clashes with top-level tmt option. `--workdir` or similar names would be confusing as that can be understood as path to run to clean.
3) supporting chaining doesn't IMHO add much value. I see two most common use-cases:
a) you want to clean everything tmt-related, you can use `tmt clean`
b) you want to clean something specific, e.g. all left-over guests; or workdirs when you are running out of space.
Cleaning exactly 2 of the three options would be nice but won't IMHO be commonly used.

Does this make sense or did we miss something? I'd be happy to turn `path` from an argument to an option if we can agree on a reasonable naming for the option but I can't think of anything clear...

**Comment by lukaszachy** [comment]
Re chaining - My use case is to clear runs and guests but keep testcloud images (bandwith/time constraints for downloads).
So `tmt clean` is a no-go and I will need to run `tmt clean guests` && `tmt clean runs` each time.

**Comment by lukaszachy** [comment]
Interrupted run seems to be a tmt issue as it doesn't note created VM anywhere: https://github.com/psss/tmt/issues/700

File: tmt/steps/provision/testcloud.py
Comment (psss) [reply tmt/steps/provision/testcloud.py:239]: For the class methods there could be used the `Class._opt()` method to get the options from the stored command line context. Did not find a better way than to have a separate method for this.

**Comment by psss** [comment]
> supporting chaining doesn't IMHO add much value. I see two most common use-cases:

I would agree with @FrNecas here. I don't see such a great difference between running those two separate commands. These commands are (probably) not something one runs many times per day so I'm fine with this little discomfort.

>  Personally I like true messages as `Would remove` more even though it is more work.

Agreed, the `would do something` wording sounds clear and true. @adka1408 added a similar changes recently:

```
> tmt init -t base --dry
Tree '/tmp/init' would be initialized.
Applying template 'base'.
Directory '/tmp/init/tests/example' would be created.
Test metadata '/tmp/init/tests/example/main.fmf' would be created.
Test script '/tmp/init/tests/example/test.sh' would be created.
Directory '/tmp/init/plans' would be created.
Plan '/tmp/init/plans/example.fmf' would be created.
```

Perhaps we could/should do a similar change for the common `run()` and `write()` methods as well.

**Comment by FrNecas** [comment]
@psss agreed, lets try to use clear conditional messages (such as `would`) in dry-mode wherever possible from now on. `run --dry` has more things to be improved (see https://github.com/psss/tmt/issues/552) other than wording..

**Comment by FrNecas** [comment]
Rebased onto latest master (there was a merge conflict). I also added some documentation similarly to `tmt status`. One slight issue I noticed is the output in case of help, for example:

```
$ tmt clean runs --help
tmt cleanup
Usage: tmt clean runs [OPTIONS] [PATH]

  Clean workdirs of past runs.

  Remove all runs in /var/tmp/tmt by default. Path to where runs should be
  searched can be specified using PATH argument.

Options:
  -l, --last          Clean the workdir of the last run.
  -i, --id ID         Run id (name or directory path) to clean workdir of.
  -k, --keep INTEGER  The number of latest workdirs to keep, clean the rest
  -v, --verbose       Show more details. Use multiple times to raise
                      verbosity.

  -d, --debug         Provide debugging information. Repeat to see more
                      details.

  -q, --quiet         Be quiet. Exit code is just enough for me.
  -n, --dry           Run in dry mode. No changes, please.
  --help              Show this message and exit.
```
Notice the first line of the output... I fixed this in https://github.com/psss/tmt/pull/646/commits/8a7d318d9da01cd1f7f415c1e8928d7ed721fd90 however it's a bit of an ugly hack but I couldn't find anything better while keeping the current formatting... Any suggestions are welcome.

File: tmt/cli.py
Comment (psss) [inline tmt/cli.py:900]:
```suggestion
        echo(style('clean', fg='red'))
```

Don't think it's necessary to mention `tmt`, everything tmt is doing is related to `tmt` ;-). What about using `clean` to be consistent with the command name?

File: tmt/cli.py
Comment (psss) [inline tmt/cli.py:962]: Brainstorm: Would it make sense to use `-h --how` as an option to provide the requested provision method instead of defining options for each method separatele? Could be a bit more extensible for the future? Not sure...

File: tmt/cli.py
Comment (psss) [inline tmt/cli.py:905]: As an alternative, instead of creating the artificial hierarchy here, we could possibly introduce a real hierarchy `Clean --> CleanRuns` and use the parent/child relation there. This would also allow calling `CleanRuns.opt('verbose')` and similar. However, I'm not convinced/sure that this rather significant change would be worth the time. Maybe in the future we can polish this?

File: tmt/utils.py
Comment (psss) [inline tmt/utils.py:531]: Defining the `load_run()` method in the `Common` class means that every object in the [hierarchy](https://tmt.readthedocs.io/en/latest/contribute.html#class-structure) will inherit it, although it is relevant for `Status` and `Clean` only. Could we find a better place for it?

File: tmt/utils.py
Comment (psss) [inline tmt/utils.py:522]: Please, add a docstring.

File: tmt/utils.py
Comment (psss) [inline tmt/utils.py:1091]: I'm not sure about the `id_ and abs_path != id_` condition. Could you please explain?

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1554]: Perhaps this line (and the duplicate a few lines above) could be under a `finally` clause?

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1559]: In `tmt run` we use all lowercase headings. What about using just `guests` here to give consistent feeling? The `clean` part is logged already so I'd propose to omit it.

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1567]: What about using `successfull` as in the method above?

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1604]: A leftover?

**Review by psss — CHANGES_REQUESTED** [review]
Thanks a lot for working on this. This must have been a bunch work! Looks good in general and works nicely. Added a couple of comments and questions.

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1504]: Would be nice to add the new class to the [overview](https://tmt.readthedocs.io/en/latest/contribute.html#class-structure).

File: tmt/cli.py
Comment (FrNecas) [reply tmt/cli.py:913]: Good point, completely forgot :) Also +1 on keeping the logging as it is now and perhaps lets brainstorm a bit more about classmethods once this PR is merged.

File: tmt/cli.py
Comment (FrNecas) [reply tmt/cli.py:962]: I was just following our initial brainstorm with this one. But your proposal probably makes more sense and will be more flexible, I will change it :+1:

File: tmt/cli.py
Comment (FrNecas) [reply tmt/cli.py:905]: Agreed, this is more of a cosmetic/design change. Lets remember this similarly to `classmethod` and logging issue

File: tmt/utils.py
Comment (FrNecas) [reply tmt/utils.py:531]: I will transform it into a function in `utils.py` rather than a method. It won't log inside the function but `Status` and `Clean` can take care of the warning, that won't be a big deal (only one logging command duplicated)

File: tmt/utils.py
Comment (FrNecas) [reply tmt/utils.py:1091]: This was basically copied over from `Status` and it probably deserves a comment in the code. The reasoning is: `id_` can be `None` (no filtering based on id). In this case, `abs_path` is "valid" (it matches the id it is filtered with). However when `id_` is not None, `abs_path` is valid only if it is equal to the specified `id_` (after conversion to absolute path).

**Comment by FrNecas** [comment]
Rebased onto the latest master (merge conflicts due to code styling fixes).

I added the improvements in new commits to make it easier to re-review @psss . Once this PR is ready to be merged, I will completely rebase this branch and somehow group the changes so that the history is a bit cleaner ;-) I made the messages a bit more consistent, I hope I didn't miss anything...

File: tmt/utils.py
Comment (psss) [reply tmt/utils.py:1100]: Thanks for the clarification.

**Review by psss — APPROVED** [review]
Thanks for all the changes and fixes. Looks good now. Ready to be merged after squash.

**Comment by psss** [comment]
Everything's green. @FrNecas, will you please squash the commits? Thanks.

**Comment by FrNecas** [comment]
Squashed, waiting for tests to pass, will merge once it's ready

**Pull request merged**
